strawberryfields/backends/states.py

```diff
@@ -311,6 +312,56 @@ def poly_quad_expectation(self, A, d=None, k=0, phi=0, **kwargs):
         """
         raise NotImplementedError
 
+    def number_expectation(self, modes):
+        r"""
+        Calculates the expectation value of the product of the number operators of the modes.
+
+        This method computes the analytic expectation value
+        :math:`\langle \hat{n}_{i_0} \hat{n}_{i_1}\dots \hat{n}_{i_{N-1}}\rangle`
+        for a (sub)set of modes :math:`[i_0, i_1, \dots, i_{N-1}]` of the system.
+
+        Args:
+            modes (list): list of modes for which one wants the expectation of the product of their number operator.
+
+        Return:
+            (float): the expectation value.
+
+        **Example**
+
+        Consider the following program:
+
+        .. code-block:: python
+
+            prog = sf.Program(3)
+
+            with prog.context as q:
+                ops.Sgate(0.5) | q[0]
+                ops.Sgate(0.5) | q[1]
+                ops.Sgate(0.5) | q[2]
+                ops.BSgate(np.pi/3, 0.1) |  (q[0], q[1])
+                ops.BSgate(np.pi/3, 0.1) |  (q[1], q[2])
+
+        Executing this on the Fock backend,
+
+        >>> eng = sf.Engine("fock", backend_options={"cutoff_dim": 10})
+        >>> state = eng.run(prog).state
+
+        we can compute the expectation value :math:`\langle \hat{n}_0\hat{n}_2\rangle`:
+
+        >>> state.number_expectation([0, 2])
+        0.07252895071309405
+
+        Executing the same program on the Gaussian backend,
+
+        >>> eng = sf.Engine("gaussian")
+        >>> state = eng.run(prog).state
+        >>> state.number_expectation([0, 2])
+        0.07566984755267293
```

Comment:
Might be worth mentioning after the example that the slight difference in value is due to the Fock basis truncation when using the Fock backend:
```suggestion
        0.07566984755267293
        
        This slight difference in value compared to the result from the Fock backend above
        is due to the finite Fock basis truncation when using the Fock backend; this can be
        avoided by increasing the value of ``cutoff_dim``.
```